src/ripple/shamap/README.md

```diff
@@ -26,164 +45,322 @@ There are two different ways of building and using a SHAMap:
+SHAMap's sequence numbers normally start out as 1.  However when a snap-shot of
+a SHAMap is made, the copy's sequence number is 1 greater than the original.
 
-It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
-snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might
-be modified must be copied before they are placed in the mutable map.
+The nodes of a SHAMap have their own copy of a sequence number.  If the SHAMap
+is not immutable, meaning it can change, then all of its nodes must have the
+same sequence number as the SHAMap itself.  This enforces an invariant that none
+of the nodes are shared with other SHAMaps.
 
+When a SHAMap gets a copy of a node from elsewhere, and must have a private
+copy, it clones the node and sets the nodes sequence number equal to the SHAMap
+sequence number.
 
-## SHAMap Thread Safety ##
+When a SHAMap decides that it is safe to share a node of its own, it sets the
+node's sequence number to 0 (a SHAMap never has a sequence number of 0). This
+is done for every node in the trie when `SHAMap::walkSubTree` is executed.
 
-*This description is obsolete and needs to be rewritten.*
+Note that other objects in rippled also have sequence numbers (e.g. ledgers).
+The SHAMap and node sequence numbers should not be confused with these other
+sequence numbers (no relation).
 
-SHAMaps can be thread safe, depending on how they are used.  The SHAMap
-uses a SyncUnorderedMap for its storage.  The SyncUnorderedMap has three
-thread-safe methods:
-
- * size(),
- * canonicalize(), and
- * retrieve()
+## SHAMap Creation ##
 
-As long as the SHAMap uses only those three interfaces to its storage
-(the mTNByID variable [which stands for Tree Node by ID]) the SHAMap is
-thread safe.
+A SHAMap is usually not created from vacuum.  Once an initial SHAMap is
+constructed, later SHAMaps are usually created by calling snapShot(bool
+isMutable) on the original SHAMap().  The returned SHAMap has the expected
+characteristics (mutable or immutable) based on the passed in flag.
 
+It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
+snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might be
+modified must be copied before they are placed in the mutable map.
+
+A new SHAMap is created with each new ledger round.  Transactions not executed
+in the previous ledger populate the SHAMap for the new ledger.
+
+## Storing SHAMap data in the database ##
+
+When consensus is reached, the ledger is closed.  As part of this process, the
+SHAMap is stored to the database by calling `SHAMap::flushDirty` with parameters
+which specify which nodes are to be written.
+
+Both `unshare()` and `flushDirty` walk the SHAMap by calling
+`SHAMap::walkSubTree`.  As `unshare()` walks the trie, nodes are not written to
+the database, and as `flushDirty` walks the trie nodes are written to the
+database. `walkSubTree` visits every node in the trie. This process must ensure
```

Comment:
I am a bit confused why `walkSubTree` would visit every node in the trie. My understanding is nodes that have not been modified do not need to be written to the database again. I was under the impression that `walkSubTree` just walked the dirty branches (the branches leading to modified/deleted/inserted nodes).

Reply:
One ledger -> one SHAMap (actually one for transactions and one for account states).  When the ledger closes, the entire SHAMap is written to the database.  I had a sentence in there that said only some nodes are written at this time.  I've removed that sentence.  It now reads:

> When consensus is reached, the ledger is closed.  As part of this process, the
> SHAMap is stored to the database by calling `SHAMap::flushDirty`.